### PR TITLE
Support multiple database for DPU

### PR DESCRIPTION
**Pull request by ganglyu**
#### Why I did it
We have updated database configuration to support multiple databases for smartswitch.
And then we need to update sonic-gnmi to support new database configuration and new GNMI path for smartswitch.
HLD PR: https://github.com/sonic-net/SONiC/pull/1625

#### How I did it
* Support new database configuration

```
{
    "INCLUDES" : [
        {
            "include" : "../../redis/sonic-db/database_config.json"
        },
        {
            "container_name" : "dpu0",
            "include" : "../../redisdpu0/sonic-db/database_config.json"
        }
    ],
    "VERSION" : "1.0"
}
```

Use a single layer instance for each device to support multiple asic and multiple dpu.
For example, for multiple asic, the instances can be named localhost, asic0, asic1, and so on. For multiple dpu, the instances can be named localhost, dpu0, dpu1, and so on. For multiple asic npu and multiple dpu, the instances can be named localhost, asic0, asic1, ..., dpu0, dpu1, and so on.

```
        "DPU_APPL_DB" : {
            "id" : 15,
            "separator": ":",
            "instance" : "redis",
            "format": "proto"
        },
```

* Update GNMI path to support multiple database

I add a new element in GNMI path to support database instance.
```
/APPL_DB/dpu0/DASH_VNET_TABLE/vnet1/
```
The second element is used for target database instance, and it can be localhost, asic0, asic1, ..., dpu0, dpu1, and so on.

#### How to verify it
Run unit test for sonic-gnmi.
Run end to end test for dash and gnmi.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

### CODE REVIEW

**Review by liuh-80 — DISMISSED** [review]

File: sonic_data_client/mixed_db_client.go [hunk 1]
```diff
@@ -26,11 +26,13 @@ import (
...
 	ssc "github.com/sonic-net/sonic-gnmi/sonic_service_client"
 	gnmipb "github.com/openconfig/gnmi/proto/gnmi"
+	"github.com/go-redis/redis"
 )
 
 const REDIS_SOCK string = "/var/run/redis/redis.sock"
 const APPL_DB int = 0
-const APPL_DB_NAME string = "APPL_DB"
+// New database for DASH
+const APPL_DB_NAME string = "DPU_APPL_DB"
```
Comment (qiluo-msft) [inline sonic_data_client/mixed_db_client.go:35]: >APPL_DB_NAME [](http://example.com/codeflow?start=6&length=12)

Disruptive change. Suggest add new constant named DPU_APPL_DB_NAME.

File: sonic_data_client/mixed_db_client.go [hunk 2]
```diff
@@ -89,13 +99,49 @@ func getMixedDbClient(zmqAddress string) (MixedDbClient) {
...
 	}
 
 	return client
 }
 
+// This function get target present in GNMI Request and
+// returns: Is Db valid (bool)
+func IsTargetDbByDBKey(dbName string, dbkey swsscommon.SonicDBKey) bool {
+	// In multiple namespaces mode, check namespace
+	// In multiple devices mode, check device
```
Comment (qiluo-msft) [inline sonic_data_client/mixed_db_client.go:113]: >multiple devices [](http://example.com/codeflow?start=7&length=16)

difficult to understand 'multiple devices mode'. do you have design doc?

File: sonic_data_client/mixed_db_client.go [hunk 3]
```diff
@@ -220,15 +243,132 @@ func (c *MixedDbClient) DbDelTable(table string, key string) error {
...
+				return "", nil, status.Error(codes.Unimplemented, "No valid elem")
+			}
+			if c.namespace_cnt > 1 || c.container_cnt > 1 {
+				if len(elems) < 2 {
+					return "", nil, status.Error(codes.Unimplemented, "Invalid elem length")
+				}
+			}
+			// Get target from the first element
+			if target == "" {
+				target = elems[0].GetName()
```
Comment (qiluo-msft) [inline sonic_data_client/mixed_db_client.go:275]: >0 [](http://example.com/codeflow?start=19&length=1)

Index 0 or 1 are magic numbers. Possible to use named constants?

File: sonic_data_client/mixed_db_client.go [hunk 4]
```diff
@@ -269,21 +417,25 @@ func (c *MixedDbClient) gnmiFullPath(prefix, path *gnmipb.Path) *gnmipb.Path {
...
 		}
-		// Skip first elem
-		fullPath.Elem = elems[1:]
+		if c.namespace_cnt > 1 || c.container_cnt > 1 {
+			// Skip first two elem for multi-namespace and multi-container
+			fullPath.Elem = elems[2:]
+		} else {
+			// Support old GNMI path schema and new GNMI path schema
+			// For device only have 1 namespace and 1 container
+			// Old GNMI path schema, /CONFIG_DB/localhost/PORT
```
Comment (qiluo-msft) [inline sonic_data_client/mixed_db_client.go:431]: >Old GNMI path [](http://example.com/codeflow?start=6&length=13)

Offline discuss backward-compatibility.

File: sonic_db_config/db_config.go [hunk 5]
```diff
@@ -194,6 +195,175 @@ func GetDbTcpAddr(db_name string, ns string) (addr string, err error) {
 	return hostname + ":" + strconv.Itoa(port), err
 }
 
+func GetDbDefaultInstance() (dbkey swsscommon.SonicDBKey, err error) {
```
Comment (qiluo-msft) [inline sonic_db_config/db_config.go:198]: >GetDbDefaultInstance [](http://example.com/codeflow?start=5&length=20)

Not used, what is the value?

File: sonic_data_client/mixed_db_client.go [hunk 6]
```diff
@@ -220,15 +243,132 @@ func (c *MixedDbClient) DbDelTable(table string, key string) error {
...
+				}
+			}
+			// Get target from the first element
+			if target == "" {
+				target = elems[0].GetName()
+			} else if target != elems[0].GetName() {
+				return "", nil, status.Error(codes.Unimplemented, "Target conflict in path")
+			}
+			if c.namespace_cnt > 1 {
+				// Get namespace from the second element
```
Comment (qiluo-msft) [inline sonic_data_client/mixed_db_client.go:280]: >namespace [](http://example.com/codeflow?start=11&length=9)

Are you assuming namespace and container could not used together? This assumption may be not necessary.

File: sonic_db_config/db_config.go [hunk 7]
```diff
@@ -194,6 +195,175 @@ func GetDbTcpAddr(db_name string, ns string) (addr string, err error) {
...
+	}
+	return dbkey_list, err
+}
+
+func GetDbInstanceFromTarget(ns string, container string) (dbkey swsscommon.SonicDBKey, ret bool) {
+	dbkey_list, err := GetDbAllInstances()
+	if err != nil {
+		return nil, false
+	}
+	for _, dbkey := range dbkey_list {
```
Comment (qiluo-msft) [inline sonic_db_config/db_config.go:263]: >for [](http://example.com/codeflow?start=1&length=3)

O(n) time complexity. Possible to improve?

File: sonic_data_client/mixed_db_client.go (hunk 1 above)
Comment (ganglyu) [reply sonic_data_client/mixed_db_client.go]: Fixed.

File: sonic_data_client/mixed_db_client.go (hunk 2 above)
Comment (ganglyu) [reply sonic_data_client/mixed_db_client.go]: Fixed, and HLD PR is https://github.com/sonic-net/SONiC/pull/1625

File: sonic_data_client/mixed_db_client.go (hunk 3 above)
Comment (ganglyu) [reply sonic_data_client/mixed_db_client.go]: Fixed

File: sonic_data_client/mixed_db_client.go (hunk 4 above)
Comment (ganglyu) [reply sonic_data_client/mixed_db_client.go]: Fixed

File: sonic_db_config/db_config.go (hunk 5 above)
Comment (ganglyu) [reply sonic_db_config/db_config.go]: Removed

File: sonic_db_config/db_config.go (hunk 7 above)
Comment (ganglyu) [reply sonic_db_config/db_config.go]: Fixed

File: sonic_data_client/mixed_db_client.go (hunk 2 above)
Comment (qiluo-msft) [reply sonic_data_client/mixed_db_client.go]: Thanks for improving comment! How about SmartSwitch with a multi-asic NPU? will your design still work?

File: sonic_data_client/mixed_db_client.go (hunk 6 above)
Comment (qiluo-msft) [reply sonic_data_client/mixed_db_client.go:286]: @zegan Could you check "sonic-swss-common does not support namespace and container together"? If yes, seems a bug.

Comment (Pterosaur) [reply sonic_data_client/mixed_db_client.go:286]: I feel the API design of sonic-swss-common should support namespace and container together.
But right now, we don't have this scenario with multi DPU + multi ASIC. So, I feel the implementation may have some issues. Not only swss-common but also other components cannot support this mixed mode.

So, IMHO, the interface design should support this mixed mode, but the implementation isn't necessary.

File: sonic_data_client/mixed_db_client.go (hunk 2 above)
Comment (ganglyu) [reply sonic_data_client/mixed_db_client.go]: I have updated, now we can support SmartSwitch with a multi-asic NPU.

File: sonic_data_client/mixed_db_client.go (hunk 6 above)
Comment (ganglyu) [reply sonic_data_client/mixed_db_client.go:308]: I have updated, now we can support SmartSwitch with a multi-asic NPU.

File: sonic_data_client/mixed_db_client.go [hunk 8]
```diff
@@ -269,21 +435,14 @@ func (c *MixedDbClient) gnmiFullPath(prefix, path *gnmipb.Path) *gnmipb.Path {
...
-	}
 	if path.GetElem() != nil {
 		elems := path.GetElem()
 		if prefix != nil {
 			elems = append(prefix.GetElem(), elems...)
 		}
-		// Skip first elem
-		fullPath.Elem = elems[1:]
+		// Skip first two elem
+		// New GNMI path schema, /CONFIG_DB/localhost/PORT
```
Comment (qiluo-msft) [inline sonic_data_client/mixed_db_client.go:444]: >localhost [](http://example.com/codeflow?start=38&length=9)

This is more general design change, not directly related to multiple database. Let's use a separate PR.

File: sonic_data_client/mixed_db_client.go (hunk 2 above)
Comment (qiluo-msft) [reply sonic_data_client/mixed_db_client.go]: closed.

File: sonic_data_client/mixed_db_client.go (hunk 6 above)
Comment (qiluo-msft) [reply sonic_data_client/mixed_db_client.go]: closed.

File: sonic_db_config/db_config.go [hunk 9]
```diff
@@ -194,6 +195,166 @@ func GetDbTcpAddr(db_name string, ns string) (addr string, err error) {
...
+	// Translate from vector to array
+	length := int(db_vec.Size())
+	for i := 0; i < length; i += 1 {
+		dbname := db_vec.Get(i)
+		db_list = append(db_list, dbname)
+	}
+	return db_list, err
+}
+
+func GetDbSeparatorByDBKey(db_name string, dbkey swsscommon.SonicDBKey) (separator string, err error) {
```
Comment (qiluo-msft) [inline sonic_db_config/db_config.go:280]: >GetDbSeparatorByDBKey [](http://example.com/codeflow?start=5&length=21)

Many functions below are 1:1 mapping of swsscommon functions. Why not directly using swsscommon functions? What is the value of the extra wrapper?

File: sonic_data_client/mixed_db_client.go (hunk 8 above)
Comment (ganglyu) [reply sonic_data_client/mixed_db_client.go]: Done

File: sonic_db_config/db_config.go (hunk 9 above)
Comment (ganglyu) [reply sonic_db_config/db_config.go:280]: This wrapper can translate swsscommon exception to error and invoke swsscommon initialize API.

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**